From 74894235044b9988bffd73c5ec61c1b55d372c51 Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Tue, 9 May 2023 10:21:57 +0200
Subject: [PATCH 1/5] Add a 'net' phase for loading user-data

We need load user-data separately before running main phase but after
network is up. Introduce a 'net' boot stage, similar to what cloud-init
does.

ref: https://cloudinit.readthedocs.io/en/latest/explanation/boot.html#network
ref: https://gitlab.alpinelinux.org/alpine/cloud/tiny-cloud/-/merge_requests/53#note_306147
---
 dist/openrc/tiny-cloud     |  2 +-
 dist/openrc/tiny-cloud-net | 14 ++++++++++++++
 lib/tiny-cloud/init        |  9 ++++++++-
 sbin/tiny-cloud            |  5 +++--
 tests/tiny-cloud.test      | 20 +++++++++++++++-----
 5 files changed, 41 insertions(+), 9 deletions(-)
 create mode 100755 dist/openrc/tiny-cloud-net

diff --git a/dist/openrc/tiny-cloud b/dist/openrc/tiny-cloud
index a4beae1..9ffdb15 100755
--- a/dist/openrc/tiny-cloud
+++ b/dist/openrc/tiny-cloud
@@ -5,7 +5,7 @@ description="Tiny Cloud Bootstrap - main phase"
 extra_commands="complete incomplete"
 
 depend() {
-	need net
+	need net tiny-cloud-net
 	before sshd
 }
 
diff --git a/dist/openrc/tiny-cloud-net b/dist/openrc/tiny-cloud-net
new file mode 100755
index 0000000..5aa8c9d
--- /dev/null
+++ b/dist/openrc/tiny-cloud-net
@@ -0,0 +1,14 @@
+#!/sbin/openrc-run
+# vim:set ts=8 noet ft=sh:
+
+description="Tiny Cloud Bootstrap - net phase"
+depend() {
+	need net
+	before *
+}
+
+start() {
+	ebegin "Tiny Cloud - net"
+	tiny-cloud net
+	eend $?
+}
diff --git a/lib/tiny-cloud/init b/lib/tiny-cloud/init
index db5cf84..8f37744 100644
--- a/lib/tiny-cloud/init
+++ b/lib/tiny-cloud/init
@@ -17,14 +17,17 @@ DEFAULT_ACTIONS_EARLY="
     create_default_user
     enable_sshd
 "
-DEFAULT_ACTIONS_MAIN="
+DEFAULT_ACTIONS_NET="
     save_userdata
+"
+DEFAULT_ACTIONS_MAIN="
     set_hostname
     set_ssh_keys
 "
 DEFAULT_ACTIONS_FINAL=""
 
 : "${INIT_ACTIONS_EARLY=$DEFAULT_ACTIONS_EARLY}"
+: "${INIT_ACTIONS_NET=$DEFAULT_ACTIONS_NET}"
 : "${INIT_ACTIONS_MAIN=$DEFAULT_ACTIONS_MAIN}"
 : "${INIT_ACTIONS_FINAL=$DEFAULT_ACTIONS_FINAL}"
 
@@ -242,6 +245,10 @@ init__set_ssh_keys() {
 
 init__save_userdata() {
     local userdata="$TINY_CLOUD_VAR/user-data"
+    if [ -f "$userdata" ]; then
+        log info "user-data already saved"
+        return
+    fi
     local tmpfile=$(mktemp "$userdata.XXXXXX")
 
     imds -e @userdata > "$tmpfile"
diff --git a/sbin/tiny-cloud b/sbin/tiny-cloud
index 4a960fa..bef2937 100755
--- a/sbin/tiny-cloud
+++ b/sbin/tiny-cloud
@@ -10,7 +10,7 @@ set -e
 
 usage() {
     cat <<EOF
-Usage: ${0##*/} [-h | --help] { early | main | final | --bootstrap {complete|incomplete} | --setup }
+Usage: ${0##*/} [-h | --help] { early | net | main | final | --bootstrap {complete|incomplete} | --setup }
 EOF
 }
 
@@ -60,7 +60,7 @@ phase="$1"
 shift
 
 case "$phase" in
-    early|main|final) ;;
+    early|net|main|final) ;;
     *) usage >&2; exit 1;;
 esac
 
@@ -92,6 +92,7 @@ INIT_ACTIONS_FINAL="${INIT_ACTIONS_FINAL} bootstrap_complete"
 
 case "$phase" in
     early)  INIT_ACTIONS="$INIT_ACTIONS_EARLY";;
+    net) INIT_ACTIONS="$INIT_ACTIONS_NET";;
     main)   INIT_ACTIONS="$INIT_ACTIONS_MAIN";;
     final)  INIT_ACTIONS="$INIT_ACTIONS_FINAL";;
     *)      usage >&2; exit 1
diff --git a/tests/tiny-cloud.test b/tests/tiny-cloud.test
index 9e0d4c8..46b1f53 100755
--- a/tests/tiny-cloud.test
+++ b/tests/tiny-cloud.test
@@ -4,11 +4,12 @@
 
 export PREFIX="$srcdir"
 export MOCK=echo
-PROVIDERS="alpine aws azure gcp nocloud oci"
+PROVIDERS="aws azure gcp nocloud oci"
 
 init_tests \
 	tiny_cloud_help \
 	no_metadata_early \
+	no_userdata_net \
 	no_userdata_main \
 	no_userdata_final
 
@@ -27,8 +28,8 @@ tiny_cloud_help_body() {
 no_metadata_early_body() {
 	fake_netcat
 	for provider in $PROVIDERS; do
-		# TODO:	-e not-match:"UNKNOWN" \
 		CLOUD="$provider" atf_check \
+			-e not-match:"UNKNOWN" \
 			-e not-match:"not found" \
 			-e not-match:"o such file" \
 			-o match:"rc-update add.*sshd" \
@@ -36,14 +37,23 @@ no_metadata_early_body() {
 	done
 }
 
+no_userdata_net_body() {
+	fake_netcat
+	for provider in $PROVIDERS; do
+		CLOUD="$provider" atf_check \
+			-e not-match:"UNKNOWN" \
+			-e match:"save_userdata.*DONE" \
+			tiny-cloud net
+	done
+}
+
 no_userdata_main_body() {
 	fake_netcat
 	for provider in $PROVIDERS; do
 		# we should not set empty hostname
 		# we should not create .ssh dir for non-existing user
-		# TODO:	-e not-match:"UNKNOWN" \
 		CLOUD="$provider" atf_check \
-			-e ignore \
+			-e not-match:"UNKNOWN" \
 			-o not-match:"hostname.*-F" \
 			-o not-match:"chown.*/\.ssh" \
 			tiny-cloud main
@@ -58,8 +68,8 @@ no_userdata_main_body() {
 no_userdata_final_body() {
 	fake_netcat
 	for provider in $PROVIDERS; do
-		# TODO:	-e not-match:"UNKNOWN" \
 		CLOUD="$provider" atf_check \
+			-e not-match:"UNKNOWN" \
 			-e match:"bootstrap_complete .*" \
 			tiny-cloud final
 		CLOUD="$provider" atf_check \
-- 
2.40.1

